Client & Server Encoding & Decoding for the STUN Binding Method

Dzol opened the pull request:
### What does this do?

An end-to-end test by dialing out to one of Google's STUN instances sending it a binding request and receiving a binding response over UDP [here](https://github.com/esl/jerboa/blob/suggest-format-module/test/jerboa_test.exs#L7). There's also encoding and decoding to service STUN binding requests.

### Checklist

This is a non exhaustive list of things which need to be done some of which are in PR #9. **Priorities**:

- [ ] change name of `%Format{}` struct and move to another module (e.g. to `%Parameters{}`)
- [x] move exceptions into their own file
- [x] exhaustive test suite (keep 100% coverage)
- [x] write exception callbacks
- [x] move exception/error modules to one place
- [x] handle not enough in binary for attributes
- [x] strip padding
- [x] stylistic consistency which tools don't catch (like those around binaries and function clauses)
- [x] get rid of the MappedAddress attribute if we don't wanna support this
- [x] raise when encoding and decoding
- [x] complete the XORMappedAddresses (IP 6 & IP 4)
- [x] check comprehension required attributes
- [x] exceptions for distinguishing elements of the STUN wire format
- [x] default values for struct fields
- [x] dialyzer annotations
- [x] decode only as much as the header says
- [x] formatting (especially the test suite)
- [x] cherry-pick Arek's commits in `master`
- [x] tags to exclude/include testing against external services
- [x] all magic cookie values in it's module
- [x] change exception/error names like [this](https://github.com/christopheradams/elixir_style_guide#exceptions)
- [x] consider removing name/value pairs in attributes

NEVER consider throwing the exceptions instead of propagating them
LATER handle too much or too little when decoding (TCP)
ELSEWHERE check port numbers are permissible
LATER test the binding scenario over TCP
LATER introduce behaviours once things are well understood
LATER macros to simplify mirroring like [this](https://github.com/esl/jerboa/blob/suggest-format-module/lib/jerboa/format/head/type.ex#L9-L17)

### Thoughts

These are points for discussion:

- raising will let us bail-out (return early) as soon as we can't encode or decode which is in keeping with the Erlang philosophy (this is **wrong**: it's not like throwing in Erlang)
- so far this PR does just encoding and decoding. We should perform other checks like permissible method and class combinations, not-well-know port ranges, duplicate attributes, and the order of attributes elsewhere (as it has nothing to do with encoding/decoding)
- the MappedAddress attribute is comprehension required but we don't support it

Commit:
Add license to repository

Commit:
Adjust style guidelines in .credo.exs

michalwski committed:
run dialyzer only for Erlang/OTP 19.2

michalwski committed:
run builds only for master branch

Commit:
Use Elixir v1.4.0

Commit:
Introduce enough to send a binding request and recieve a response

arkgil commented:
Hey, it's great it works against Google's server! I suggest that we mark tests interacting with external services with `external` tag, and configure ExUnit to not run these every time we run a test suite. And configure Travis to run `mix test --include external`.

arkgil commented:
Regarding
> raising will let us bail-out (return early) as soon as we can't encode or decode which is in keeping with the Erlang philosophy

I think that we can introduce bang (!) versions of encode and decode functions, and raise our custom exception if we encounter an error.

Commit:
Improve documentation and follow some stylistic conventions

Ignore some of the suggestions if they don't seem reasonable.

Commit:
Remove support for Mapped Address attribute

This attribute is present in STUN RFC only for compatibility
with older STUN clients.

Commit:
Remove IPAddress struct and put it into XORMappedAddress module

Commit:
Introduce decode exceptions and wrap values with :ok or :error

Elixir exceptions give us a consistent and general way to encapsulate
failure information which we send to a caller who propagates these so
that we can ultimately send a STUN failure response with this
information. There are some other minor changes: e.g. introduce module
attributes where appropriate.

Joseph + Arek.

Commit:
Introduce & test decoding IPv6 XORMappedAddress attributes

Commit:
Improve XORMappedAddress test's binary construction

Move attribute values into functions which can be put in a helper
module in the future.

Commit:
Introduce :error tuple for unknown comprehension required attribute

Commit:
Introduce & propagate exceptions if input is not STUN wire format

These are constructed when we don't see the distinguishing elements of
the STUN wire format in the header's leading two bits and an the
lengths trailing two bits.

Commit:
Test out and squash identifier decoding bug

Commit:
Refactor and add new exceptions

Commit:
Add property tests for STUN header format

Commit:
Test decoding too short and too long packets

Commit:
Make tests asynchronous

Commit:
Put given, when, then-like parts of cases on different lines

Commit:
Default to [] for the `attributes' list in the Format struct

Commit:
Introduce types and specs for API functions & re-work documentation

Joseph + Arek.

Commit:
Exclude system tests by default but include them on Travis

This is so that we don't run them every time when we test locally.

Commit:
Improve exception (names & messages) and follow conventions

Some names are generalized while others are made more explicit. All
exceptions end with "Error". We keep messages to a single sentence and
start with a lowercase letter.

Commit:
Move logic about body length (as given in header) to format module

The header gives us a length for the body, but it's niether the
header's concern nor the body's to make sure it's correct, rather it's
the whole message's concern. So move it from the body's decoding logic
to the `Format` module which is responsible for everything.

Commit:
Format binary/bit-syntax in a consistent way with no white-space

They get too long otherwise.

Commit:
Simplify bit-syntax for encoding the Type into the header

Commit:
Rename `head` to `header` as named in the STUN RFC

Commit:
Change field name for bytes following the end of a STUN message

Thinking about it, `extra` makes more sense, as `excess` suggests it's
a continuation of the message that was already there (when, in fact,
it isn't).

Commit:
Strip attributes of padding

Commit:
Move exceptions into the Format module to make names shorter

They're distinguishable enough for this to be okay and it makes sense
to put them here because this is where we will deal with them (as they
propagate up to here).

Commit:
Add property tests for XORMappedAddress decoding

Commit:
Ignore optional attributes + test attribute errors propagate to body

Commit:
Strip padding off of the end of an attribute (not beginning)

Commit:
Replace message/1 w/ exception/1 in exceptions

Commit:
Test the decode!/1 API

Commit:
Apply style suggestions from Credo

Commit:
Fix return type of Attribute.decode

Commit:
Build branch for basic encoding & decoding on Travis

We're going to force this history to master.

Commit:
Test we can decode/encode header for a server-side binding scenario

I.e. the server has to decode a binding request and encode a binding
response.

Commit:
Build and test encoding the XORMappedAddress attribute

Commit:
Build & test encoding the IPv6 XORMappedAddress attribute

Commit:
Introduce encoding attributes (along w/ XORMA) into TLV binaries

TLV: Type-Length-Value
XORMA: XORMappedAddress

Commit:
Dramatically improve clarity of XORMappedAddress encoding tests

Intoroduce a helper module and default into the binary building
helpers.

Commit:
Introduce & test encoding just one attribute into a body

Commit:
Build & test encoding complete STUN messages with a header and a body

Commit:
Neaten XORMappedAddress decoding and formatting elsewhere

Commit:
Move Format prefixed exceptions into their own file

After writig encoding/1 in the Format module it was obviously annoying
to have the exceptions listed before the module's main functions.

Commit:
Improve XORMA encoding tests by introducing assertion helpers

Commit:
Improve attribute encoding tests by introducing assertion helpers

Commit:
Improve body encoding tests by introducing assertion helper

Commit:
Introduce test helper which totals numbers in keyword list

This helps make testing binary sizes more descriptive. See the commit
for how nice it is.

Commit:
Safeguard a little in case we try to encode to big an attribute

Commit:
Merge pull request #11 from esl/server-side-binding

Server Encoding & Decoding for the STUN Binding Method

Commit:
Fix warnings about types unknown to dialyzer

Modules containing types weren't aliased properly.

Commit:
Fix dialyzer warnings about exception/1 callbacks

arkgil commented:
Pushed to master as a foundation of work on Jerboa. Closing
